src/starkware_utils/commitment_tree/calculation.rs

```diff
@@ -0,0 +1,276 @@
+use std::any::Any;
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use crate::starkware_utils::commitment_tree::binary_fact_tree::BinaryFactDict;
+use crate::starkware_utils::commitment_tree::errors::CombineError;
+use crate::starkware_utils::commitment_tree::leaf_fact::LeafFact;
+use crate::starkware_utils::commitment_tree::patricia_tree::nodes::PatriciaNodeFact;
+use crate::storage::storage::{FactFetchingContext, HashFunctionType, Storage};
+
+// These enums are used instead of trait objects because the conditions to turn
+// LeafFact and InnerNodeFact into object safe traits are complex to lift: multiple places where
+// `Sized` is required, async methods, generic type parameters, etc.
+// This can be solved somehow by moving some features to `Storage`, but is out of scope for now.
+// These enums do the trick in the meantime, at the cost of a little flexibility.
+
+pub type CalculationInnerNodeFact = PatriciaNodeFact;
+
+/// A mapping between a hash and its corresponding fact. Split into two maps, one for leaves and
+/// one for inner nodes.
+pub struct NodeFactDict<LF> {
+    pub inner_nodes: HashMap<Vec<u8>, CalculationInnerNodeFact>,
+    pub leaves: HashMap<Vec<u8>, LF>,
+}
+
+impl<LF> Default for NodeFactDict<LF> {
+    fn default() -> Self {
+        Self { inner_nodes: Default::default(), leaves: Default::default() }
+    }
+}
+
+/// A calculation that can produce a result of type T. The calculation is dependent on the results
+/// of other calculations. Those calculations can be of type other than T.
+/// The result of the calculation can be produced when the results of the dependency calculations
+/// are given.
+pub trait Calculation<T, LF> {
```

Comment:
Seems like this is used primarily for hashing, it might help to call that example out in the documentation (or rename the trait if that makes sense)

Reply:
In order:
* I'll make it `pub(crate)`, sure.
* In terms of documentation, I'll make a separate PR at some point. I'll try to prepare a workshop for next Thursday so this might be part of it.
* I won't rename the trait, I'd like to keep it close to the Python implementation to make it easier to track changes in future versions.